cli/src/commands/extensions/permissions.rs

```diff
@@ -61,6 +62,65 @@ impl Permission {
             Err(anyhow!("Requires {resource_type} access to {resource:?}"))
         }
     }
+
+    pub fn subset_of(&self, parent: &Permission) -> Result<Permission> {
+        match (parent, self) {
+            // Child deny-all always succeeds, returning deny-all.
+            (_, &Permission::Boolean(false)) => Ok(Permission::Boolean(false)),
+            // Parent deny-all fails with all child permissions but deny-all.
+            (&Permission::Boolean(false), _) => {
+                Err(anyhow!("All permissions are denied by the manifest"))
+            },
+            // Parent allow-all always succeeds, returning the child's permissions.
+            (&Permission::Boolean(true), child) => Ok(child.clone()),
+            // Child allow-all fails with more restrictive parent permissions.
+            (_, &Permission::Boolean(true)) => {
+                Err(anyhow!("The requested permissions are denied by the manifest"))
+            },
+            // Parent set vs child set have to be validated.
+            // This will error if child is not subset of parent, and return the child set otherwise.
+            (&Permission::List(ref parent), &Permission::List(ref child)) => {
+                Permission::check_paths_include_children(parent, child)
+                    .map(|_| Permission::List(child.clone()))
+                    .map_err(|mismatches| {
+                        anyhow!(
+                            "The following paths are denied by the manifest: {}",
+                            mismatches.join(", ")
+                        )
+                    })
+            },
+        }
+    }
+
+    fn check_paths_include_children(
+        parent: &[String],
+        child: &[String],
+    ) -> StdResult<(), Vec<String>> {
+        let parent_paths = parent.iter().map(PathBuf::from).collect::<Vec<_>>();
+        let child_paths = child.iter().map(PathBuf::from).collect::<Vec<_>>();
+
+        // Find all paths in `child` that don't have a prefix in `parent`.
+        let without_parent: Vec<_> = child_paths
+            .iter()
+            .filter_map(|child| {
+                // Using PathBuf::starts_with rather than String::starts_with in order to get
+                // the correct semantics.
+                if parent_paths.iter().any(|p| child.starts_with(p)) {
```

Comment:
The `starts_with` function is in `Path` and not `PathBuf`. Which means that you could probably refactor this function to avoid extra allocations and copies.

For a simple option, if `c` and `p` are both `&String`, then you can call `Path::starts_with` like this: `Path::new(c).starts_with(p)`

Reply:
Usually these sorts of optimizations are automatically tackled by LLVM and generate the same code; out of curiosity, I benchmarked these two functions and both consistently take ~170ns per iteration:

```rust
pub fn with_path() -> bool {
    let parent = "/tmp".to_string();
    let child = "/tmp/child".to_string();

    let parent_path = Path::new(&parent);
    let child_path = Path::new(&child);

    child_path.starts_with(parent_path)
}

pub fn with_pathbuf() -> bool {
    let parent = "/tmp".to_string();
    let child = "/tmp/child".to_string();

    let parent_path = PathBuf::from(parent);
    let child_path = PathBuf::from(child);

    child_path.starts_with(parent_path)
}
```

The change is minuscule, literally switching `PathBuf::from` for `Path::new` at the top of the method, so I think it's worth it anyway just in case.

Reply:
I also should have been more clear: Using `Path` means you can completely avoid building the `parent_paths` and `child_paths` vectors. I'll add a suggestion on the new code to show this better.

Reply:
Here's the same thing with `String` in lieu of `&str`: https://godbolt.org/z/Kzse56ohe

Interestingly enough, `with_path` calls `drop_in_place<PathBuf>`.

Reply:
> Interestingly enough, `with_path` calls `drop_in_place<PathBuf>`.

For what it's worth, I think that's actually `drop_in_place<String>` (which gets collapsed into `drop_in_place<PathBuf>` during optimization). And it's only called in the unwind section (i.e., during a panic)